### ISSUE TITLE
geocode: v2 cascade router — single decision tree compiled to branchless Rust

### ISSUE THREAD

**Issue by pierre-warnier** [butterfly-osm/butterfly-osm#226]
Sub-issue of #225 (world-resilient architecture). Replaces the byte-transformer-as-primary with a learned decision tree compiled to branchless Rust at build time.

## Decisions

1. **Entropy is the feature filter.** Generous candidate feature set (200-400 features). Tree training picks via information-gain. No hand-curated short list. Useless features get zero-gain leaves, pruned naturally.
2. **Skip bench panels for now.** Ship v2 first, bench after. Continental coverage (#225 sub-c) follows.
3. **Single decision tree** (not ensemble GBDT). Simpler, faster, sufficient. Escalate to small GBDT only if error analysis shows interaction failures.
4. **Build-time codegen.** \`build.rs\` reads trained tree JSON at compile time, emits inlined Rust source.

## Architecture

```
butterfly-geocode train-country-classifier --corpus ... --out classifier.json
                                                                    |
                                                            (committed to repo)
                                                                    |
                                              build.rs reads JSON at compile time
                                                                    |
                                              walks tree, emits OUT_DIR/country_classifier.rs
                                                  as nested if/else or match-on-feature-idx
                                                                    |
                                              include!'d into geocode crate
                                                                    |
                                              compiler inlines + branch-eliminates
                                                                    |
                                              runtime: features [f32; ~50] -> top-K posterior
                                              ~10-50 ns inference
```

## Feature spec (generous, entropy-filtered)

Single byte-pass extraction (~3-5 µs total), accumulates ~200-400 candidate features:

- **Script flags + ratios** per UTF-8 script: Latin, CJK, Cyrillic, Arabic, Devanagari, Hangul, Thai, Hebrew, Greek (~20 features)
- **Digit features**: total digit count, longest digit run, runs of length 2/3/4/5/6/7+, presence of \\d{4}, \\d{5}, \\d{7}, \\d-\\d block patterns (~15 features)
- **Lexical marker presences**: every street/place marker from every country pack morphology TOML (\`Straße\`, \`Rua\`, \`丁目\`, \`ул\`, \`نهج\`, \`Av\`, \`Boulevard\`, \`Blvd\`, etc.) — ~80-150 features pulled from existing 15 morphology packs
- **Punctuation features**: comma, period, hyphen, parens, slash counts; CJK punctuation (\`、\`, \`。\`); RTL marker presence (~12 features)
- **Length features**: byte count, char count (UTF-8 codepoints), token count, avg token length (~5 features)
- **Casing features**: title-case ratio, all-caps ratio, mixed-case ratio (CJK: zero), no-case fraction (~6 features)
- **Diacritic features**: presence of à é ñ ç ü ö ł ž č etc. (~10 features)
- **N-grams**: top 50-100 most-frequent byte 2-grams from training corpus (each as 0/1 presence) (~50-100 features)
- **Mixed-script flag** + multi-script ratio (~3 features)

The training data picks which of these matter. Family-aware emerges from data — if 'longest digit run' splits well at threshold 6.5 for distinguishing JP-block from US-Western, the tree learns it.

## Implementation plan

Single PR, multi-commit, all phases:

1. **Phase 1**: Feature extraction module \`geocode/src/routing/features.rs\` — one byte-pass struct, all candidate features. Unit tests with hand-crafted inputs (BE/JP/RU/AR/Hindi samples).

2. **Phase 2**: Training CLI \`butterfly-geocode train-country-classifier --corpus ... --out classifier.json\`. Reads gold corpora, extracts features, trains via \`gbdt\` crate (\`n_estimators=1, max_depth=12\`), serializes to JSON. Pure Rust.

3. **Phase 3**: \`geocode/build.rs\` — reads \`geocode/data/models/country-classifier.json\` at compile time, walks tree, generates \`OUT_DIR/country_classifier.rs\`. Generated code: \`pub fn classify(features: &Features) -> [(CountryId, f32); K]\` with inlined comparisons.

4. **Phase 4**: Replace existing \`geocode/src/routing/classifier.rs\` invocation. Server \`/geocode\` handler: cascade outputs top-K country posterior → recall queries top-K country FSTs in parallel → rerank consolidates.

5. **Phase 5**: Train initial classifier on existing 15-country gold corpora. Commit \`country-classifier.json\` to repo. Verify generated code compiles + matches model on held-out test set.

6. **Phase 6**: Bench v2 on Belgium 1k-query TSV vs current pipeline. Quick smoke. Continental panels per #225 sub-c is separate follow-up.

## Acceptance

- \`cargo build --release\` produces inlined classifier code
- Inference latency p50 ≤ 100 ns measured (10× margin under target)
- Classifier accuracy on held-out OA test set ≥ 95% top-1 country
- BE recall@1 stays at 39% (current baseline) — this PR is plumbing + speed, not accuracy gain. Accuracy improvements come via #225 sub-(b) augmentation + iterative feature additions.
- Workspace clippy + fmt + tests stay green

## What this is NOT

- Not a Belgium fix
- Not a multi-country tagger replacement (yet — that comes later if v2 stabilizes)
- Not bench coverage (separate, #225 sub-c)
- Not an accuracy gain in itself; it's a throughput + architecture clarity gain

## Dependencies

None. Builds on main HEAD. Ships in one PR.

**Comment by pierre-warnier** [comment]
## Measurement results from #227 — recommendation locked in

**Use a tree** (15 one-vs-rest gbdt, depth 12, 1 estimator each). Empirically beats signature-table by 70pp and cost-aware-cascade by 60pp at the relevant feature scale.

### Knee point
- **356 features**, **624 ns extraction**, **738 ns p50 inference**, **870 KB model**
- Macro top-1: **88.88%** on the 15-country gold corpus
- Feature classes that survive: `marker + digit + bigram + length + punct`
- Feature classes to DROP: script (+0.02% only — corpus is Latin-only), postcode (+0.4% but +1670 ns extraction cost — digit class already captures the discriminative info)

### Weak points to address before / during #226 ship
1. **BE 69.7% / CH 74.9% top-1** — multilingual border ambiguity. Concrete fix: add postcode-prefix bins as features (`postcode_be_1xxx`, `be_2xxx`, etc.). Not measured in #227.
2. **No non-Latin samples in training corpus.** Macro 88.88% is on Latin-only data; western_latin family is the weakest at 86.71% which is the headline. Cyrillic/Arabic/Hangul/Thai/Hebrew untested.

### Hard gate before #226 ships

**Block on #225 sub-(c) continental bench panels.** The world-resilience claim cannot be validated on the current corpus. We need bench data covering CJK / Cyrillic / Arabic / Devanagari / Hangul / Thai / RTL families with per-family acceptance gates ≥50% top-1.

### What ships in #226 once unblocked

- Tree training CLI (`butterfly-geocode train-country-classifier`) — pure Rust, gbdt crate
- build.rs codegen: JSON tree → inlined Rust source
- Top-K country posterior fed to multi-FST recall in parallel
- ~5-10 µs end-to-end (feature extraction dominates; tree inference is ~50 ns when compiled)

Reference: PR #228 ships the measurement code. `geocode-research/router-features/` contains the 391-feature extractor, training/curve/ablation CLIs, and artifacts. Reproducible.

**Comment by pierre-warnier** [comment]
Shelving the geocoder. See goodbye ticket #253 for context. State preserved in tag geocode-shelved-2026-05-23 (resurrect via 'git checkout tags/geocode-shelved-2026-05-23 -- geocode/' ).
